modules/prompt_parser.py

```diff
@@ -23,68 +55,62 @@
+    class AtStep(Transformer):
+        @v_args(tree=True)
+        def prompt(self, tree):
+            children = []
+            for child in tree.children:
+                if child.data != 'scheduled':
+                    children.append(child)
+                    continue
+                p1, p2, t = child.children[0].children, child.children[-2].children, child.children[-1]
+                children.extend(p1 if step <= t else p2)
+            tree.children = children
+            return tree
+    return AtStep().transform(tree)
 
 def get_learned_conditioning_prompt_schedules(prompts, steps):
     res = []
-    cache = {}
-
     for prompt in prompts:
-        prompt_schedule: list[list[str | int]] = [[steps, ""]]
-
-        cached = cache.get(prompt, None)
-        if cached is not None:
-            res.append(cached)
-            continue
-
-        for m in re_prompt.finditer(prompt):
-            plaintext = m.group(1) if m.group(5) is None else m.group(5)
-            concept_from = m.group(2)
-            concept_to = m.group(3)
-            if concept_to is None:
-                concept_to = concept_from
-                concept_from = ""
-            swap_position = float(m.group(4)) if m.group(4) is not None else None
-
-            if swap_position is not None:
-                if swap_position < 1:
-                    swap_position = swap_position * steps
-                swap_position = int(min(swap_position, steps))
-
-            swap_index = None
-            found_exact_index = False
-            for i in range(len(prompt_schedule)):
-                end_step = prompt_schedule[i][0]
-                prompt_schedule[i][1] += plaintext
-
-                if swap_position is not None and swap_index is None:
-                    if swap_position == end_step:
-                        swap_index = i
-                        found_exact_index = True
-
-                    if swap_position < end_step:
-                        swap_index = i
-
-            if swap_index is not None:
-                if not found_exact_index:
-                    prompt_schedule.insert(swap_index, [swap_position, prompt_schedule[swap_index][1]])
-
-                for i in range(len(prompt_schedule)):
-                    end_step = prompt_schedule[i][0]
-                    must_replace = swap_position < end_step
-
-                    prompt_schedule[i][1] += concept_to if must_replace else concept_from
-
-        res.append(prompt_schedule)
-        cache[prompt] = prompt_schedule
-        #for t in prompt_schedule:
-        #    print(t)
-
+        tree = parser.parse(prompt)
+        key_steps = collect_steps(steps, tree)
+        res.append([(k, at_step(k, tree)) for k in key_steps])
     return res
 
 
 ScheduledPromptConditioning = namedtuple("ScheduledPromptConditioning", ["end_at_step", "cond"])
 ScheduledPromptBatch = namedtuple("ScheduledPromptBatch", ["shape", "schedules"])
 
+from itertools import product

```

Comment:
Imports at the top? Or is that not the python way.

Reply:
Nah just me being lazy. Should have used a linter.